File: go/with_permit.go
package atlasent

import "context"

// WithPermit authorizes a request and runs fn only if AtlaSent issued
// and verified a permit. Returns whatever fn returns.
//
// Where [Client.Protect] returns a verified *Permit and leaves the caller
// to invoke their own action, WithPermit orchestrates the entire
// lifecycle in a single call:
//
//  1. evaluate the request → return *DeniedError on anything other than ALLOW.
//  2. verify the resulting permit → return *DeniedError on Verified=false
//     (covers v1 single-use semantics: a permit consumed by an earlier
//     verify reports verified:false on a replay).
//  3. invoke fn with the verified permit.
//  4. return fn's result.
//
// fn never runs unless steps 1 and 2 succeed. Errors returned by fn
// itself propagate verbatim — the permit is already consumed by step 2
// in v1, so the SDK can't meaningfully roll back; surfacing the
// caller's error is the right behaviour.
//
// Internally, WithPermit delegates to [Client.Protect] so the two paths
// can never drift in fail-closed semantics or error taxonomy. Future
// Protect-side hardening (risk-tier gating, etc.) flows through
// automatically.
//
//	result, err := atlasent.WithPermit(
//	    ctx,
//	    client,
//	    atlasent.EvaluateRequest{
//	        Agent:   "deploy-bot",
//	        Action:  "deploy_to_production",
//	        Context: map[string]any{"commit": commit, "approver": approver},
//	    },
//	    func(permit *atlasent.Permit) (string, error) {
//	        return doDeploy(commit, permit.PermitID)
//	    },
//	)
//
// Mirrors withPermit() (TypeScript, PR #123) and with_permit() (Python).
func WithPermit[T any](
	ctx context.Context,
	c *Client,
	req EvaluateRequest,
	fn func(*Permit) (T, error),
) (T, error) {
	var zero T
	permit, err := c.Protect(ctx, req)
	if err != nil {
		return zero, err
	}
	return fn(permit)
}

File: typescript/src/withPermit.ts
/**
 * `atlasent.withPermit(...)` — the lexically-scoped form of
 * {@link protect}. TypeScript mirror of the Python SDK's
 * ``atlasent.with_permit(...)``.
 *
 * Same execution-boundary contract as {@link protect}: evaluate +
 * verifyPermit run end-to-end before the executor is invoked, and the
 * executor cannot run unless a verified {@link Permit} was returned.
 * The difference is purely lexical — `withPermit` binds the execution
 * to the permit's lifetime via a callback, so the call site reads as
 * "execute this body under a permit" rather than "fetch a permit and
 * run code under it manually."
 *
 * ```ts
 * import atlasent from "@atlasent/sdk";
 *
 * const ok = await atlasent.withPermit(
 *   {
 *     agent: "deploy-bot",
 *     action: "production.deploy",
 *     context: { commit, approver },
 *   },
 *   async (permit) => {
 *     const result = await runDeploy(commit);
 *     return { ok: true, permitId: permit.permitId, result };
 *   },
 * );
 * ```
 *
 * Pick the form that fits the call site:
 *
 * - {@link protect} when the caller wants the verified permit as a
 *   value (e.g. to pass it across a process boundary, persist it
 *   alongside their own record, or interleave it with non-trivial
 *   control flow).
 * - {@link withPermit} when the action body is a single lexical scope
 *   and "no permit, no execution" is the only thing the call site
 *   needs to express.
 *
 * Both surfaces use the same underlying primitive and produce the
 * same audit-chain entry.
 */

import { protect, type Permit, type ProtectRequest } from "./protect.js";

/**
 * Authorize a request end-to-end and invoke `fn` only on a verified
 * permit.
 *
 * @param request Same shape as {@link ProtectRequest}.
 * @param fn Invoked with the verified {@link Permit}. Its return
 *   value (awaited if it is a promise) is propagated to the caller.
 *
 * @returns Whatever `fn` returns.
 *
 * @throws {AtlaSentDeniedError} Policy denied, hold/escalate, or
 *   permit failed verification. `fn` is never invoked.
 * @throws {AtlaSentError} Transport, timeout, auth, rate-limit, or
 *   server error. `fn` is never invoked.
 *
 * Errors thrown inside `fn` propagate untouched — the permit is
 * already consumed by the verify step in v1, so there is no
 * compensating revoke.
 */
export async function withPermit<T>(
  request: ProtectRequest,
  fn: (permit: Permit) => Promise<T> | T,
): Promise<T> {
  const permit = await protect(request);
  return await fn(permit);
}
